### PR TITLE
Improvements for nested samplers

### PR DESCRIPTION
**Pull request by aymgal**
Main changes :
* all samplers now return _equally weighted_ (aka _unweighted_) samples. This is better to visualise posteriors on current plotting routines
* all nested samplers are grouped in a single FittingSequence `'nested_sampling'` routine, and `'sampler_type'` can be `'MULTINEST'`, `'DYPOLYCHORD'` or `'DYNESTY'` (see docstrings to see settings related to each sampler)
* the `sigma_scale` parameter for those samplers now scales the parameter space widths, when `prior_type` is `'uniform'`. In any case, please tell me if you feel that it should be the same behaviour for `'gaussian'` prior as well (and not a scaling of the standard deviation as it is now)

### CODE REVIEW

**Comment by sibirrer** [comment]
Hi Aymeric,
tests are not passing yet: https://travis-ci.org/sibirrer/lenstronomy

File: lenstronomy/Workflow/fitting_sequence.py [hunk 1]
```diff
@@ -241,6 +223,66 @@ def pso(self, n_particles, n_iterations, sigma_scale=1, print_key='PSO', threadC
...
+
+        elif sampler_type == 'DYPOLYCHORD':
+            sampler = DyPolyChordSampler(self.likelihoodModule,
+                                         prior_type=prior_type,
+                                         prior_means=mean_start,
+                                         prior_sigmas=sigma_start,
+                                         sigma_scale=sigma_scale,
+                                         output_dir='dypolychord_chains',
+                                         output_basename=output_basename,
+                                         remove_output_dir=remove_output_dir)
```
Comment (ajshajib) [inline lenstronomy/Workflow/fitting_sequence.py]: Please add argument `use_mpi=self._mpi` when calling `DyPolyChordSampler()`.

File: lenstronomy/Workflow/fitting_sequence.py [hunk 2]
```diff
@@ -241,6 +223,66 @@ def pso(self, n_particles, n_iterations, sigma_scale=1, print_key='PSO', threadC
...
+                = sampler.run(dypolychord_dynamic_goal, kwargs_run)
+
+        elif sampler_type == 'DYNESTY':
+            sampler = DynestySampler(self.likelihoodModule,
+                                     prior_type=prior_type,
+                                     prior_means=mean_start,
+                                     prior_sigmas=sigma_start,
+                                     sigma_scale=sigma_scale,
+                                     bound=dynesty_bound, 
+                                     sample=dynesty_sample)
```
Comment (ajshajib) [inline lenstronomy/Workflow/fitting_sequence.py]: Please add arguments `use_mpi=self._mpi, use_pool=use_pool` when calling `DynestySampler()`.

File: lenstronomy/Workflow/fitting_sequence.py [hunk 3]
```diff
@@ -241,6 +223,66 @@ def pso(self, n_particles, n_iterations, sigma_scale=1, print_key='PSO', threadC
...
+        if sampler_type == 'MULTINEST':
+            sampler = MultiNestSampler(self.likelihoodModule,
+                                       prior_type=prior_type,
+                                       prior_means=mean_start,
+                                       prior_sigmas=sigma_start,
+                                       sigma_scale=sigma_scale,
+                                       output_dir='multinest_chains',
+                                       output_basename=output_basename,
+                                       remove_output_dir=remove_output_dir,
+                                       use_mpi=False)
```
Comment (ajshajib) [inline lenstronomy/Workflow/fitting_sequence.py]: I haven't tested MPI with MultiNest, may be your plan is to test it later, or I can test it sometime down the line. I was not planning to use it anytime soon. However, I am not sure if you are turning off MPI with hard-coding on purpose here. I guess, once MPI is properly tested to work with MultiNest, it should be `use_mpi=self._mpi`.

File: lenstronomy/Workflow/fitting_sequence.py [hunk 4]
```diff
@@ -241,6 +223,66 @@ def pso(self, n_particles, n_iterations, sigma_scale=1, print_key='PSO', threadC
...
+
+        elif sampler_type == 'DYNESTY':
+            sampler = DynestySampler(self.likelihoodModule,
+                                     prior_type=prior_type,
+                                     prior_means=mean_start,
+                                     prior_sigmas=sigma_start,
+                                     sigma_scale=sigma_scale,
+                                     bound=dynesty_bound, 
+                                     sample=dynesty_sample)
+            samples, means, logZ, logZ_err, logL = sampler.run(kwargs_run)
```
Comment (ajshajib) [inline lenstronomy/Workflow/fitting_sequence.py]: Would it be possible to return the `results` object from the dynesty sampler? It has some useful functions for quickly checking the result. Is there any downside to returning it?

File: lenstronomy/Workflow/fitting_sequence.py (hunk 3 above)
Comment (aymgal) [reply lenstronomy/Workflow/fitting_sequence.py]: I don't plan to test it very soon, but indeed, this option should be available from the FittingSequence interface. As for MCMC / PSO, I will pass the `self._mpi`

File: lenstronomy/Workflow/fitting_sequence.py (hunk 4 above)
Comment (aymgal) [reply lenstronomy/Workflow/fitting_sequence.py]: Yes it will be done in the next commits

**Comment by aymgal** [comment]
Ok implemented the changes, let's hope that tests will go well

**Comment by sibirrer** [comment]
Hi Aymeric, some tests do still not pass as the samplers probably can not get installed on travis properly or that the test exception was not complete. Are you able to build a work-around such that the tests are passing?
https://travis-ci.org/sibirrer/lenstronomy/jobs/553786889
Thank you!

**Comment by aymgal** [comment]
With those last few commits, tests are now running fine, and coverage should be ok : https://travis-ci.org/sibirrer/lenstronomy/builds/554220730

**Comment by sibirrer** [comment]
Thank you very much! I know it's a bit of a hassle sometimes with travis and external libraries.

**Pull request merged**